### ISSUE TITLE
exp: track contacts-v1 R-precision and validation loss over time

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#180]
## Kind

- Kind: evals

## Question

How good was the best contacts-v1 model we had on any given date — on contact
R-precision and on held-out validation loss — and how tightly do those two
track each other?

Every number exists somewhere already (#67, #75, #82, #89, #108, #117, #120,
#146, #160, #169), but they are scattered across issue comments, per-experiment
CSVs and two W&B projects, with two different inference recipes mixed in. This
experiment collects them into one dataset and three standing figures, and keeps
them refreshable as new models land.

## Hypothesis

Not a hypothesis-testing experiment — it is a **standing tracker**. The two
things it is built to make visible, both of which prior issues assert
piecemeal:

1. Progress on contact accuracy has come almost entirely from the **base
   model**, not from inference or post-training.
2. Validation loss predicts contact accuracy **across training generations**
   and stops predicting it **within a generation** (#169's finding), so the
   loss frontier and the accuracy frontier are not the same curve.

## Background

- **#89** — the fixed 554-protein contact benchmark: ground-truth universe,
  candidate pairs, and `compute_metrics.py`. Every R-precision number here is
  computed by that implementation.
- **#82** — settled the best LM-only inference: n=100 rollouts + per-rollout
  document resampling + pairwise tie-break. **#142** then removed the
  `top_k=50` that was truncating rollouts.
- **#75 / #117 / #146** — Eric's LR/WD/epoch sweeps (`eric-czech/marin`, tags
  `exp75`, `exp117`, `exp146`, `exp153`); **#67 / #85 / #108 / #120 / #137 /
  #150 / #155** — MarinFold-side runs (`open-athena/MarinFold`).
- **#169** — showed val loss is not a usable checkpoint selector at the
  0.01-nat scale, and that matched loss does not mean matched accuracy across
  model sizes. This experiment is the longitudinal version of its panel A.
- **#74 / #78** — the structure-predictor baselines drawn as reference lines.

## Approach

Two tables and three figures, all code in the experiment dir.

- `build_dataset.py` — pulls every finished contacts-v1 training run reporting
  `eval/contacts-v1-val/loss` from both W&B projects, and carries the
  hand-curated benchmark scores (one source citation per row, since these live
  in issue comments and per-experiment CSVs rather than in W&B). Recomputes the
  structure-predictor lines from exp89's per-protein table.
- `plot_progress.py` — the three figures: R-precision vs date, val loss vs
  date, R-precision vs val loss. Both date figures draw a running-best
  staircase; both R-precision figures carry the structure predictors as dotted
  reference lines.

**The one methodological trap, handled explicitly:** the same weights score
**~0.086 higher** under exp82's rollout recipe than under exp89's original
pairwise scorer (#61/#75 E8: 0.339 → 0.425; #120: 0.350 → 0.436). The figures
keep the two recipes visually separate rather than pooling them.

Metric throughout: **R-precision, all ranges (seq-sep ≥ 6), mean over the 554
proteins**. Date = when the training run finished, not when it was evaluated.

## Success criteria

- One dataset with a traceable source for every plotted number, and three
  figures that a future reader can regenerate with two commands.
- A documented refresh procedure so the tracker stays current as #146, #150,
  #153, #155 and the post-training line produce new checkpoints.

## Notes

Follow-up worth doing separately: **#108's 3B on CoreWeave held the loss
frontier from 2026-07-11 to 07-16 at 2.7418 and was never contact-scored.**
Given #146 (a 3B at 0.0012 better loss scoring 0.023 worse on R-precision) it
is probably not a missed frontier point, but that is an inference, not a
measurement, and it is the only real gap in the accuracy figure.

**Comment by timodonnell** [comment]
🤖 **Built and populated.** Code, data and figures in **PR #181** · experiment dir [`experiments/exp180_evals_contacts_v1_progress_over_time/`](https://github.com/Open-Athena/MarinFold/tree/claude/rprecision-validation-loss-plots-750306/experiments/exp180_evals_contacts_v1_progress_over_time)

![R-precision over time](https://github.com/Open-Athena/MarinFold/blob/claude/rprecision-validation-loss-plots-750306/experiments/exp180_evals_contacts_v1_progress_over_time/plots/rprecision_frontier.png?raw=true)

**The accuracy frontier moved in exactly two jumps, both from the base model:** ~0.03 → **0.425** when #75's E8 rung finished (2026-06-21), and 0.436 → **0.534** when #117's 16-epoch bs256 run finished (2026-07-22). Between them, five weeks of post-training and inference work moved it by **+0.011** (#120's re-epoch); #160's backtracking fine-tune moved it **−0.020**. Still below single-seq Protenix-v2 (0.603) and well below ESMFold2 (0.786).

The loss frontier has far more steps — 3.15 → 2.7566 (#75 E8) → **2.7418** (#108's 3B on CoreWeave, 07-11) → 2.7213 (#120) → 2.7131 → 2.7112 → 2.7037 (#117) → 2.7025 (#146), over 153 finished runs — which is the point: **loss and accuracy agree across generations and stop agreeing inside one.** ~2 R-precision per nat over the #75→#117 gap; nothing at all over #117's early-vs-final 0.008 nats; and #146's 3B is 0.0012 *better* on loss and 0.023 *worse* on R-precision.

Both hypotheses in the issue body hold as stated.

## Two corrections to existing records

Neither is changed in the PR — flagging rather than editing history unilaterally.

1. **`UPDATES.md`, week of June 22** says #85's LR re-heat "improved eval loss somewhat". It did not: it finished at **2.9801** against #67's **2.9800**, and its five eval points run 2.9825 / 2.9828 / 2.9843 / 2.9820 / 2.9801. Happy to amend in a follow-up.
2. **#67 never held the loss frontier.** It finished 2026-06-14 15:36 at 2.9800, about two hours after `prot-exp75-cv1-1_5b-e2-lr7e-4-wd0p05-v1` reached 2.9787.

## The gap worth closing

**#108's 3B on CoreWeave held the loss frontier from 07-11 to 07-16 at 2.7418 and was never contact-scored.** Given #146 — a 3B at 0.0012 better loss scoring 0.023 worse — it is probably not a missed frontier point, but that is an inference and it is the only real hole in the accuracy figure. One ~10-min benchmark run would settle it.

## Keeping it current

The README documents the refresh procedure. Two commands (`build_dataset.py` re-pulls W&B, `plot_progress.py` redraws) for new training runs; a hand-added row in `RPRECISION_ROWS` for a new benchmark score. The trap it is written around: **the same weights score ~0.086 higher under exp82's rollout recipe than under exp89's pairwise scorer**, so a score filed under the wrong recipe would look like a jump that never happened. Anything from the `eval-checkpoint` skill is pairwise; anything from exp82's rollout workers or the exp169 dispatcher is rollout.

Leaving this issue open — it is a standing tracker, and closure is a human call either way.

**Comment by timodonnell** [comment]
🤖 **Audit: are any published eval numbers affected by the transformers-5 rope defect?**

**Conclusion: no. Every published contact number stands.** Details below, including what is verified versus inferred.

## The test each eval has to pass

A checkpoint is at risk only if **both** hold: it was exported by transformers 5.x (so `config.json` states rope as `rope_parameters`), **and** it was scored by a stack pinned to transformers 4.x (which ignores that key and substitutes the architecture default).

Affected exports, from `scripts/repair_checkpoint_config.py --survey`:

| repo / path | exporter | at risk |
|---|---|---|
| `marinfold-exp117` · bs256 step-35679 | 5.12.1 | yes |
| `marinfold-exp117` · bs64 step-71359 | 5.12.1 | yes |
| `marinfold-exp146` · 3B e8 step-17839 | 5.12.1 | yes |
| `marinfold-exp146` · 3B e2 step-17839 | 5.12.1 | yes |
| `marinfold-exp75` · step-35679 | 4.57.5 | no |
| bucket: exp120, exp67, exp75 | 4.x | no |

## Eval by eval

| eval | checkpoints | verdict | basis |
|---|---|---|---|
| **#89** (#67, #75 E1/E2/E4/E8, ens10) | 4.x exports | ✅ clean | configs carry `rope_theta` natively; nothing to lose |
| **#82** rollout, #75 E8 = 0.4245 | 4.x export | ✅ clean | same |
| **#82** rollout, #117 = 0.5350 | 5.x export | ✅ clean | **verified by agreement** — exp169 scored the same checkpoint from an independently repaired export and got 0.5344. A 0.0006 gap is backend noise; a dropped rope is worth 0.77 nats |
| **#169** (#117 final/early, #146 3B) | 5.x exports | ✅ clean | **verified in code** — `prepare_hf_export.py` splits `rope_parameters` into `rope_theta`+`rope_scaling`, and `verify_prepared_exports.py` *hard-fails* if `rope_scaling` isn't llama3 or if `rope_parameters` survives |
| **#120** arms (exp89 pairwise) | 4.x exports | ✅ clean | native |
| **#160** (exp120-base, bt50) | 4.x exports | ✅ clean | native; bt50's 0.4158 is also in the consistent range |
| **Eric's gist** exp117-E8-bs64 = 0.4192 | 5.x export | ✅ clean | **inferred** — his stack is transformers 5.x end to end (it produced the export), so it reads `rope_parameters` natively. The defect is a 4.x-*consumer* problem, not a property of the checkpoint |

## Why a subtle bias was never on the table

This is the part that makes the audit cheap: **a wrong rope base is catastrophic, not marginal.** Loading #117 at theta 10000 instead of 500000 costs **0.765 nats/token** (2.414 → 3.179, measured on three real contacts-v1 documents). For calibration, E1–E4 in #75 — genuinely undertrained models — score R-precision ≈ 0.03 against E8's 0.339.

So an affected eval would not read slightly low. It would read *near chance*. Every published number instead moves monotonically with val loss and in the expected direction:

```
#67 0.029 → #75 E8 0.339 → #120 0.350 → #117 E8 0.419   (pairwise)
        #75 E8 0.425 → #120 0.436 → #117 E16 0.534       (rollout)
```

There is no room in that progression for a broken load.

## What I did not do

I did not re-score any checkpoint to confirm. Given the effect size, re-scoring would only re-derive numbers we already have; if anyone wants belt-and-braces on the one **inferred** row (Eric's gist), the cheap version is re-running that single checkpoint through the #89 harness from the repaired export and checking it lands near 0.419.

## Fixes now in flight

- **Upstream, so it stops recurring:** [marin-community/marin#7801](https://github.com/marin-community/marin/pull/7801) — levanter's HF export now writes both rope shapes. Additive, no-op under 5.x, verified on transformers 4.57 and 5.14. This is what closes it for #146, #150, #155 and #160 exports going forward.
- **Our loaders:** #184 repairs on load, so the CLI and eval harness are correct regardless of which exporter wrote a checkpoint.
- **The artifact:** #184 also republishes #117 under our bucket with a config both transformers majors read, so a bare `AutoModelForCausalLM.from_pretrained` is correct with no MarinFold code in the loop. Eric's repos were left untouched.

**Still carrying the defect and deliberately not modified:** `marinfold-exp146`'s two exports and exp117's bs64 export. They are only a hazard for a 4.x consumer, and `--survey` reports them read-only.

**Comment by timodonnell** [comment]
Tracker refreshed with #199 / #204 (PR #207). **The frontier moved twice in two days.**

![Best contacts-v1 model trained to date — contact R-precision](https://raw.githubusercontent.com/Open-Athena/MarinFold/04b9ba8d15166eccf5aa1d4ea2e9e462c2ef8990/experiments/exp180_evals_contacts_v1_progress_over_time/plots/rprecision_frontier.png)

| date | model | R-precision (all) | val loss |
|---|---|---:|---:|
| 2026-07-31 | #166 AA aug | 0.562 | 2.6642 |
| **2026-08-09** | **#199 TRC p03-base** | **0.578** | ~2.6257 |
| **2026-08-10** | **#199 CoreWeave p06-aug** | **0.587** | ~2.5895 |

Also plotted but not steps: #199 TRC p03-aug 0.574 (folded into the p03-base label, see below) and TRC p06-aug 0.524.

**The single-sequence Protenix-v2 gap has effectively closed.** 0.587 against 0.603 is a paired Δ of **−0.016, 95% CI [−0.041, +0.009]** over the 554 proteins. Protenix still has the higher mean — this is not a claim to have passed it, it is a claim that this benchmark can no longer separate them. Two months ago the same comparison read 0.029 vs 0.603. ESMFold2 (0.786) and Protenix+MSA (0.812) are untouched, and that is where the remaining distance is.

Every step of this staircase is still a base-model result. Inference and post-training have never taken it.

### Two things that are new, beyond the numbers

**A noise floor.** #204 evaluated the unchanged #117 checkpoint four times: they span **0.0023**. That is the first such estimate this tracker has had — every earlier "X beat Y by 0.00n" here was a between-run subtraction with no yardstick at all. It makes #199's p03-base vs p03-aug difference (0.0036) a tie, so the two now share one label rather than drawing two steps. It does *not* cover #199 CW vs #166 (0.025) or CW vs TRC p06-aug (0.063).

**The loss→accuracy exchange rate is collapsing:**

| step | Δ loss | Δ R | R per nat |
|---|---:|---:|---:|
| #75 E8 → #117 E16 | 0.053 | +0.109 | 2.06 |
| #117 E16 → #166 | 0.040 | +0.028 | 0.71 |
| #166 → #199 CW | ~0.075 | +0.025 | ~0.33 |

#204's sigmoid fit says the same thing from the other side — upper asymptote 0.5955, which #199 CW is already at 98.6% of. Either the relationship saturates near there or the fit is extrapolating past its support; the next frontier point will tell. Either way #169's recommendation gets stronger: select on the contact metric, use loss only to decide what is worth scoring.

### Loss-axis caveat

The `~` above is load-bearing. marin [#7209](https://github.com/marin-community/marin/pull/7209) changed the packed-LM objective to mask padding targets, which raises the reported loss by **~0.38 nats**, and #199 is the first sweep recorded under it. The tracker now declares a scale per source, plots the historical axis (97% of runs are natively there) and converts the rest — drawn in green with a `~` on both loss figures.

Worth knowing for anything else that compares losses across this boundary: **the scale follows the pinned marin version, not the run date**, and it is not in the W&B config. MarinFold-side runs *launched in August* still pin a June `marin-core` and remain historical; exp199 uses `0.2.73`. I read it off each source's own `requirements.txt` artifact rather than assuming. The conversion (`current ≈ historical + 0.38171`, from [Eric's same-checkpoint gist](https://gist.github.com/eric-czech/9c40252457790a513eeb62a6a965c049)) is good to ~0.025 nats where #199 sits — fine at the 0.1-nat scale, unreadable at the 0.01-nat scale. **R-precision is unaffected.**

### Also in PR #207

- Head-to-head scatters re-pointed from #166 to #199 CW. The 200–400 length bin flips positive (+0.013 on 171 proteins), joining the >400 bin — MarinFold now wins the upper half of the length range against a single-sequence structure predictor and loses the lower half.
- The oracle best-of-100 headroom (+0.041, on #155's checkpoint) is now *larger* than the 0.016 still separating the best model from single-sequence Protenix. Rough ordering across two checkpoints, not arithmetic — but the ordering is the point.
- Recorded as caveats: #199's CoreWeave-vs-TRC gap confounds initialisation, schedule and step count so it is not a platform result; and #155 and #199 both add ESM-Atlas but were never run against each other.
